src/Nethermind/Nethermind.Trie/NibbleExtensions.cs

```diff
@@ -18,26 +21,46 @@ public static Nibble[] FromBytes(params byte[] bytes)
         public static Nibble[] FromBytes(ReadOnlySpan<byte> bytes)
         {
             Nibble[] nibbles = new Nibble[2 * bytes.Length];
-            for (int i = 0; i < bytes.Length; i++)
-            {
-                nibbles[i * 2] = new Nibble((byte)((bytes[i] & 240) >> 4));
-                nibbles[i * 2 + 1] = new Nibble((byte)(bytes[i] & 15));
-            }
-
+            BytesToNibbleBytes(bytes, MemoryMarshal.AsBytes(nibbles.AsSpan()));
             return nibbles;
         }
 
-        public static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
+        public unsafe static void BytesToNibbleBytes(ReadOnlySpan<byte> bytes, Span<byte> nibbles)
         {
             if (nibbles.Length != 2 * bytes.Length)
             {
                 ThrowArgumentException();
             }
 
-            for (int i = 0; i < bytes.Length; i++)
+            var length = bytes.Length / sizeof(Vector128<byte>) * sizeof(Vector128<byte>);
+            if (Vector128.IsHardwareAccelerated && length > 0)
+            {
+                var input = MemoryMarshal.Cast<byte, Vector128<byte>>(bytes.Slice(0, length));
+                ref var output = ref Unsafe.As<byte, Vector128<ushort>>(ref MemoryMarshal.GetReference(nibbles));
+
+                for (int i = 0; i < input.Length; i++)
+                {
+                    Vector128<byte> value = input[i];
+                    (Vector128<ushort> lower0, Vector128<ushort> upper0) = Vector128.Widen(Vector128.BitwiseAnd(value, Vector128.Create((byte)0xf)));
+                    lower0 = Vector128.Shuffle(lower0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
+                    upper0 = Vector128.Shuffle(upper0.AsByte(), Vector128.Create((byte)1, 0, 1, 2, 1, 4, 1, 6, 1, 8, 1, 10, 1, 12, 1, 14)).AsUInt16();
+                    (Vector128<ushort> lower1, Vector128<ushort> upper1) = Vector128.Widen(Vector128.BitwiseAnd(value, Vector128.Create((byte)0xf0)));
+                    lower1 = Vector128.ShiftRightLogical(lower1.AsByte(), 4).AsUInt16();
+                    upper1 = Vector128.ShiftRightLogical(upper1.AsByte(), 4).AsUInt16();
```

Comment:
Can you add some comments? explaining what it does?

Reply:
Added comments